### PR TITLE
👍 Improve codec performance

### PR DESCRIPTION
**Pull request by tkrs**
This PR improved codec performance by avoiding to lookup of the `TypeCodec` via `CodecRegistry`. 

<details>
<summary>Benchmark</summary>
- Cassandra to Scala's Future

```
[info] Benchmark                                                          Mode  Cnt        Score        Error   Units
[info] JavaDriverFutureBenchmark.one                                     thrpt   10     1817.417 ±    125.378   ops/s
[info] JavaDriverFutureBenchmark.one:·gc.alloc.rate                      thrpt   10       27.122 ±     13.187  MB/sec
[info] JavaDriverFutureBenchmark.one:·gc.alloc.rate.norm                 thrpt   10    18427.197 ±   8313.877    B/op
[info] JavaDriverFutureBenchmark.one:·gc.churn.PS_Eden_Space             thrpt   10       29.789 ±     18.075  MB/sec
[info] JavaDriverFutureBenchmark.one:·gc.churn.PS_Eden_Space.norm        thrpt   10    20824.903 ±  11787.468    B/op
[info] JavaDriverFutureBenchmark.one:·gc.churn.PS_Survivor_Space         thrpt   10        0.030 ±      0.080  MB/sec
[info] JavaDriverFutureBenchmark.one:·gc.churn.PS_Survivor_Space.norm    thrpt   10       22.703 ±     57.968    B/op
[info] JavaDriverFutureBenchmark.one:·gc.count                           thrpt   10        9.000               counts
[info] JavaDriverFutureBenchmark.one:·gc.time                            thrpt   10       29.000                   ms
[info] JavaDriverFutureBenchmark.three                                   thrpt   10      475.107 ±      7.826   ops/s
[info] JavaDriverFutureBenchmark.three:·gc.alloc.rate                    thrpt   10      394.862 ±    208.716  MB/sec
[info] JavaDriverFutureBenchmark.three:·gc.alloc.rate.norm               thrpt   10  1017705.351 ± 535712.596    B/op
[info] JavaDriverFutureBenchmark.three:·gc.churn.PS_Eden_Space           thrpt   10      427.213 ±     90.820  MB/sec
[info] JavaDriverFutureBenchmark.three:·gc.churn.PS_Eden_Space.norm      thrpt   10  1145286.277 ±  94689.517    B/op
[info] JavaDriverFutureBenchmark.three:·gc.churn.PS_Survivor_Space       thrpt   10        0.167 ±      0.122  MB/sec
[info] JavaDriverFutureBenchmark.three:·gc.churn.PS_Survivor_Space.norm  thrpt   10      467.755 ±    413.597    B/op
[info] JavaDriverFutureBenchmark.three:·gc.count                         thrpt   10       71.000               counts
[info] JavaDriverFutureBenchmark.three:·gc.time                          thrpt   10       90.000                   ms
```

- Twitter's Future

```
[info] TwitterFutureBenchmark.one                                        thrpt   10     1940.012 ±     84.747   ops/s
[info] TwitterFutureBenchmark.one:·gc.alloc.rate                         thrpt   10       30.002 ±     14.014  MB/sec
[info] TwitterFutureBenchmark.one:·gc.alloc.rate.norm                    thrpt   10    19113.456 ±   8122.955    B/op
[info] TwitterFutureBenchmark.one:·gc.churn.PS_Eden_Space                thrpt   10       32.331 ±      8.625  MB/sec
[info] TwitterFutureBenchmark.one:·gc.churn.PS_Eden_Space.norm           thrpt   10    21266.087 ±   3971.453    B/op
[info] TwitterFutureBenchmark.one:·gc.churn.PS_Survivor_Space            thrpt   10        0.029 ±      0.069  MB/sec
[info] TwitterFutureBenchmark.one:·gc.churn.PS_Survivor_Space.norm       thrpt   10       18.422 ±     43.315    B/op
[info] TwitterFutureBenchmark.one:·gc.count                              thrpt   10       10.000               counts
[info] TwitterFutureBenchmark.one:·gc.time                               thrpt   10       34.000                   ms
[info] TwitterFutureBenchmark.three                                      thrpt   10      527.581 ±     10.288   ops/s
[info] TwitterFutureBenchmark.three:·gc.alloc.rate                       thrpt   10      477.489 ±    252.156  MB/sec
[info] TwitterFutureBenchmark.three:·gc.alloc.rate.norm                  thrpt   10  1109070.145 ± 582714.759    B/op
[info] TwitterFutureBenchmark.three:·gc.churn.PS_Eden_Space              thrpt   10      517.938 ±    110.703  MB/sec
[info] TwitterFutureBenchmark.three:·gc.churn.PS_Eden_Space.norm         thrpt   10  1250470.609 ± 104846.442    B/op
[info] TwitterFutureBenchmark.three:·gc.churn.PS_Survivor_Space          thrpt   10        0.229 ±      0.142  MB/sec
[info] TwitterFutureBenchmark.three:·gc.churn.PS_Survivor_Space.norm     thrpt   10      557.359 ±    323.683    B/op
[info] TwitterFutureBenchmark.three:·gc.count                            thrpt   10       76.000               counts
[info] TwitterFutureBenchmark.three:·gc.time                             thrpt   10       91.000                   ms
```
</details>

### CODE REVIEW

File: core/src/main/scala/agni/Deserializer.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+          size match {
+            case 0 => builder.result().asRight
+            case x =>
+              val kbb = CodecUtils.readValue(input, version)
+              val vbb = CodecUtils.readValue(input, version)
+              val r = (K.apply(kbb, version) |@| V.apply(vbb, version)).map {
+                case (k, v) =>
+                  acc += k -> v
+              }
+              r match {
```
Comment (yanana) [inline core/src/main/scala/agni/Deserializer.scala:128]:
```scala
r.flatMap(_ => go(x - 1, input, acc)
```

looks better to me.

Comment (tkrs) [reply core/src/main/scala/agni/Deserializer.scala:128]: I'm afraid not. Because `flatMap` is not a tail recursive.

Comment (yanana) [reply core/src/main/scala/agni/Deserializer.scala:128]: Ah, I see. You're completely right😉

**Review by yanana — APPROVED** [review]
LGTM

**Pull request merged**